### ISSUE TITLE
autosummary and Python 3 problem

### ISSUE THREAD

**Issue by shimizukawa** [sphinx-doc/sphinx#1261]
I have a problem building a sphinx project with Python 3. The project uses autosummary and the :toctree: directive, an example file is here:

https://raw.github.com/lingpy/lingpy/master/doc/source/docu/compare/trebor.rst

When I try to build the project I get the error:

```
  File "/usr/local/lib/python3.2/dist-packages/Sphinx-1.2b1-py3.2.egg/sphinx/ext/autosummary/generate.py", line 104, in generate_autosummary_docs
    for name, path, template_name in sorted(items):
TypeError: unorderable types: NoneType() < str()
```

This seem to be an error related to the "sorted" function in Python 3. In Python 2 everything is fine. The line in "generated.py" is the following:

https://github.com/sphinx-doc/sphinx/blob/88c72be190afd98cc1619624ae37b8955f393e97/sphinx/ext/autosummary/generate.py#L104

The next line 105 indicates that some tuple entries might be "None", so sorting should also work with "None" values. Is this a bug in autosummary?

Best,
Peter

---
- Bitbucket: https://bitbucket.org/birkenfeld/sphinx/issue/1261
- Originally reported by: [pbouda](https://bitbucket.org/pbouda)
- Originally created at: 2013-09-10T10:34:43.710

**Comment by shimizukawa** [comment]
_From [pbouda](https://bitbucket.org/pbouda) on 2013-09-10 08:39:06+00:00_

Btw, we fixed this temporarily by replacing line 104 of generate.py with:

```
#!python

for name, path, template_name in sorted(items,key=lambda x: str(x)):
```

I am not sure if this is the intended behaviour, but we can then build our documentation also with Python 3.

**Comment by shimizukawa** [comment]
_From Takayuki Shimizukawa on 2013-09-10 08:47:56+00:00_

This issue can be reproduced with both 1.1.3 and 1.2b1.

**Comment by shimizukawa** [comment]
_From Georg Brandl on 2014-01-12 21:34:01+00:00_

Closes #1261: Python 3 cannot sort None and str types, make everything a str.

→ <<cset 31de96330d2f68bcc25085ecc63ea6c942b1c908>>
